x/config/configurator.go

```diff
@@ -0,0 +1,648 @@
+		return nil, fmt.Errorf("failed to decode %v: %v", b.cfgType, err)
+	}
+
+	// Note: the receiver is not on the pointer so that these objects are
+	// re-usable.
+	b.cfg = cfg.Elem()
+	return &b, nil
+}
+
+func (b *transportBuilder) BuildTransport() (transport.Transport, error) {
+	result := b.build.Call([]reflect.Value{b.cfg})
+	if err, _ := result[1].Interface().(error); err != nil {
+		return nil, err
+	}
+	return result[0].Interface().(transport.Transport), nil
+}
+
+// Builder for types that depend on a transport.Transport as their first
+// argument.
+//
+// This logic is shared between inbounds and different outbound types. It's
+// very similar to transportBuilder except they all depend on a
+// transport.Transport as their first argument.
+//
+// This implements functions to be a valid InboundBilder,
+// UnaryOutboundBuilder, and OnewayOutboundBuilder but it's up to the logic in
+// this package to ensure that this is not used incorrectly.
+type builder struct {
+	cfg        reflect.Value
+	cfgType    reflect.Type
+	resultType reflect.Type
+	buildFunc  reflect.Value
+	fieldNames map[string]struct{}
+}
+
+func newBuilder(build interface{}, outputType reflect.Type) (*builder, error) {
+	v := reflect.ValueOf(build)
+	t := v.Type()
+
+	var err error
+	switch {
+	case t.Kind() != reflect.Func:
+		err = errors.New("must be a function")
+	case t.NumIn() != 2:
+		err = fmt.Errorf("must accept exactly two arguments, found %v", t.NumIn())
+	case !isDecodable(t.In(0)):
+		err = fmt.Errorf("must accept a struct or struct pointer as its first argument, found %v", t.In(0))
+	case t.In(1) != _typeOfTransport:
+		// TODO: We can make this smarter by making transport.Transport
+		// optional and either the first or the second argument instead of
+		// requiring it as the second argument.
+		err = fmt.Errorf("must accept a transport.Transport as its second argument, found %v", t.In(1))
+	case t.NumOut() != 2:
+		err = fmt.Errorf("must return exactly two results, found %v", t.NumOut())
+	case t.Out(0) != outputType:
+		err = fmt.Errorf("must return a %v as its first result, found %v", outputType, t.Out(0))
+	case t.Out(1) != _typeOfError:
+		err = fmt.Errorf("must return a error as its second result, found %v", t.Out(1))
+	}
+
+	return &builder{
+		cfgType:    t.In(0),
+		fieldNames: fieldNames(t.In(0)),
+		resultType: outputType,
+		buildFunc:  v,
+	}, err
+}
+
+func (b builder) Load(attrs attributeMap) (*builder, error) {
+	cfg := reflect.New(b.cfgType)
+	if err := attrs.Decode(cfg.Interface()); err != nil {
+		return nil, fmt.Errorf("failed to decode %v: %v", b.cfgType, err)
+	}
+
+	// Note: the receiver is not on the pointer so that these objects are
+	// re-usable.
+	b.cfg = cfg.Elem()
+	return &b, nil
+}
+
+func (b *builder) build(t transport.Transport) (reflect.Value, error) {
+	out := b.buildFunc.Call([]reflect.Value{b.cfg, reflect.ValueOf(t)})
+	if err, _ := out[1].Interface().(error); err != nil {
+		return reflect.Zero(b.resultType), err
```

Comment:
nit: why do you care about first arg to be zero initialized? it looks like if can be omitted.